### ISSUE TITLE
Do you have the plan for draft UEFI Specification for LoongArch about relocation type in PE format(LoongArch specially)?

### ISSUE THREAD

**Issue by TYDQSoft** [loongson/la-abi-specs#9]
I have read the code in binutils for loongarch and it only provide three relocation constant.It is too few to use,I think.Do you have the plan for loongarch UEFI specification (for PE format in loongarch)?

**Comment by SixWeining** [comment]
Did you mean this https://learn.microsoft.com/en-us/windows/win32/debug/pe-format that already supports LoongArch?

![image](https://github.com/loongson/la-abi-specs/assets/90239436/723c5957-156d-43de-a517-1f6cdce1a5aa)

**Comment by SixWeining** [comment]
Cc @yetist @kilaterlee who may know more details.

**Comment by xen0n** [comment]
Context (in Chinese): https://bbs.loongarch.org/d/443

I guess maybe the OP wants more complete relocs in PE, similar to the full set of `R_LARCH_B*`, `R_LARCH_ABS*`, `R_LARCH_PCALA*` etc., so that a complete PE toolchain could be made.

I agree that the OP should provide more information regarding their motivation and background etc. so we don't have to guess as much.

**Comment by TYDQSoft** [comment]
> Did you mean this https://learn.microsoft.com/en-us/windows/win32/debug/pe-format that already supports LoongArch?
> 
> ![image](https://private-user-images.githubusercontent.com/90239436/346410797-723c5957-156d-43de-a517-1f6cdce1a5aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1MDQ4NTcsIm5iZiI6MTcyMDUwNDU1NywicGF0aCI6Ii85MDIzOTQzNi8zNDY0MTA3OTctNzIzYzU5NTctMTU2ZC00M2RlLWE1MTctMWY2Y2RjZTFhNWFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA5VDA1NTU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlYjNjZDgwMGU3ZGIzMWMwODE4MDE2YzM5NjJlNjczOTZjYjRlMzllYTYzNzUyMmI0NmM0M2E0ZTE3ZDBhOTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5YUffq8rhRZ4v-X5BOBQRgQVdml5W534GORIDhe1cVg)

However,I need more loongarch-specific relocation constants such as IMAGE_REL_LOONGARCH_ABSOLUTE.

**Comment by driver1998** [comment]
There are two constants defined in .NET PAL, not documented in PE spec (not sure if they are .NET specific):

IMAGE_REL_LOONGARCH64_PC
IMAGE_REL_LOONGARCH64_JIR

https://github.com/dotnet/runtime/blob/a5cc707d976a14495462c9c492a921ff0927b8f5/src/coreclr/pal/inc/rt/ntimage.h#L1021
implementation: https://github.com/dotnet/runtime/blob/af5e7150362294696d3d52aa2302062e4b7c8095/src/coreclr/tools/Common/Compiler/DependencyAnalysis/Relocation.cs#L542

**Comment by shushanhf** [comment]
> There are two constants defined in .NET PAL, not documented in PE spec (not sure if they are .NET specific):
> IMAGE_REL_LOONGARCH64_PC IMAGE_REL_LOONGARCH64_JIR

No, these are only used for .NET and only LoongArch64's AOT within .NET.

**Comment by TYDQSoft** [comment]
> There are two constants defined in .NET PAL, not documented in PE spec (not sure if they are .NET specific):
> 
> IMAGE_REL_LOONGARCH64_PC IMAGE_REL_LOONGARCH64_JIR
> 
> https://github.com/dotnet/runtime/blob/a5cc707d976a14495462c9c492a921ff0927b8f5/src/coreclr/pal/inc/rt/ntimage.h#L1021 implementation: https://github.com/dotnet/runtime/blob/af5e7150362294696d3d52aa2302062e4b7c8095/src/coreclr/tools/Common/Compiler/DependencyAnalysis/Relocation.cs#L542

So that's why draft UEFI specification for LoongArch about relocation type in PE format matters.I need a LoongArch64 relocation table for free pascal compiler.These .NET only constants are so disappointed for free pascal compiler users.

**Comment by driver1998** [comment]
From the code it looks like there is really only **one** relocation type supported in UEFI: IMAGE_REL_BASED_LOONGARCH64_MARK_LA.

So if you need to target UEFI, this is the one you need to use.

https://github.com/tianocore/edk2/blob/1bb9f47739ae7993191a36bea76c5a2157fdd10f/MdePkg/Library/BasePeCoffLib/LoongArch/PeCoffLoaderEx.c#L43
